prompt2model/param_selector/search_with_optuna.py

```diff
@@ -0,0 +1,245 @@
+"""This module provides automatic hyperparameter selection using optuna."""
+
+from __future__ import annotations  # noqa FI58
+
+import os
+from typing import Any, Optional
+
+import optuna
+import transformers
+from datasets import Dataset, concatenate_datasets
+from optuna.trial import Trial
+from transformers import PreTrainedModel  # noqa
+from transformers import PreTrainedTokenizer
+
+from prompt2model.model_trainer.generate import GenerationModelTrainer
+from prompt2model.param_selector.base import ParamSelector
+from prompt2model.prompt_parser.base import PromptSpec
+from prompt2model.utils.config import DEFAULT_HYPERPARAMETERS_SPACE
+
+
+class OptunaParamSelector(ParamSelector):
+    """Uses optuna for searching for hyperparameters."""
+
+    def __init__(self, trainer: GenerationModelTrainer, n_trials: int):
+        """Initializes a new instance of OptunaParamSelector.
+
+        Args:
+            trainer (BaseTrainer): trainer object from GenerationModelTrainer
+            n_trials (int): The number of trials for conducting hyperparameter search
```

Comment:
Can we elaborate a little bit on what this means? e.g. 
```suggestion
            n_trials (int): The maximum number of parameter configurations to evaluate
                during conducting hyperparameter search.
```